cmd/compute-domain-kubelet-plugin/health.go

```diff
@@ -0,0 +1,143 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+	drapb "k8s.io/kubelet/pkg/apis/dra/v1beta1"
+	registerapi "k8s.io/kubelet/pkg/apis/pluginregistration/v1"
+)
+
+type healthcheck struct {
+	grpc_health_v1.UnimplementedHealthServer
+
+	server *grpc.Server
+	wg     sync.WaitGroup
+
+	regClient registerapi.RegistrationClient
+	draClient drapb.DRAPluginClient
+}
+
+func startHealthcheck(ctx context.Context, config *Config) (*healthcheck, error) {
+	port := config.flags.healthcheckPort
+	if port < 0 {
+		return nil, nil
+	}
+
+	addr := net.JoinHostPort("", strconv.Itoa(port))
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen for healthcheck service at %s: %w", addr, err)
+	}
+
+	regSockPath := (&url.URL{
+		Scheme: "unix",
+		// TODO: this needs to adapt when seamless upgrades
+		// are enabled and the filename includes a uid.
+		Path: path.Join(DriverRegistrarPath, DriverName+"-reg.sock"),
+	}).String()
+	klog.V(6).Infof("connecting to registration socket path=%s", regSockPath)
+	regConn, err := grpc.NewClient(
+		regSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to registration socket: %w", err)
+	}
+
+	draSockPath := (&url.URL{
+		Scheme: "unix",
+		Path:   path.Join(DriverPluginPath, "dra.sock"),
+	}).String()
+	klog.V(6).Infof("connecting to DRA socket path=%s", draSockPath)
+	draConn, err := grpc.NewClient(
+		draSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to DRA socket: %w", err)
+	}
+
+	server := grpc.NewServer()
+	healthcheck := &healthcheck{
+		server:    server,
+		regClient: registerapi.NewRegistrationClient(regConn),
+		draClient: drapb.NewDRAPluginClient(draConn),
+	}
+	grpc_health_v1.RegisterHealthServer(server, healthcheck)
+
+	healthcheck.wg.Add(1)
+	go func() {
+		defer healthcheck.wg.Done()
+		klog.Infof("starting healthcheck service at %s", lis.Addr().String())
+		if err := server.Serve(lis); err != nil {
+			klog.Errorf("failed to serve healthcheck service on %s: %v", addr, err)
+		}
+	}()
+
+	return healthcheck, nil
+}
+
+func (h *healthcheck) Stop() {
+	klog.Info("Stopping healthcheck service")
+	if h.server != nil {
+		h.server.GracefulStop()
+	}
+	h.wg.Wait()
+}
```

Comment:
The original PR has this log inside the `if` statement. I'M fine differing here if we think its useful, but having it insid the if means that it will only get printed if the server has actually been started first.